### ISSUE TITLE
Can't do full-colour image previews under Fedora / W3MIMGDISPLAY_PATH environment variable does not work correctly

### ISSUE THREAD

**Issue by lycurgus** [hut/ranger#449]
I've encountered the same problem that's documented in #104 - Fedora places the `w3mimgdisplay` binary under `/usr/libexec/w3m/` instead of the `/usr/lib/w3m/` that ranger expects. I have set the `W3MIMGDISPLAY_PATH` environment variable to the correct path, but it doesn't seem to be used.

I've done some digging and if I alter the `W3MIMGDISPLAY_PATH` variable in `ranger/ext/img_display.py` to point at the correct path, everything works as intended. Adding a couple of lines to log the contents of the `path` variable (I'm on Fedora 23, which has ranger v1.6.1 currently, so it hasn't changed to `self.binary_path`) seems to indicate that if the environment variable is set, its value is read - but unless the path is correct in the Python variable, an error message ("no such file or directory") is shown in the ranger window when an attempt is made to preview an image.

It seems therefore that ranger is reading the environment variable but not using it. I'm happy to either symlink the binary to the expected location or leave the altered variable in `img_display.py` - but it seems like the better solution would be to ensure that the environment variable is respected.

**Comment by hut** [comment]
How did you change the environment variable? Maybe you do it in a place that is ignored in your particular setup.

**Comment by lycurgus** [comment]
The environment variable is set in my .bashrc - if I do an `echo $W3MIMGDISPLAY` at a bash prompt, it shows the path correctly. I altered `img_display.py` to log what it believes the binary path is, and if the environment variable is set it logs that (correct) path, but still fails to execute it.

So far the scenarios I have tried are:

* environment variable set to correct path, `img_display.py` unaltered -> fails to run
* environment variable set to correct path, `img_display.py` altered to also have the correct path -> runs as expected

In both cases, logging the value of the `path` variable shows the correct path - so it seems the environment variable is being read successfully - but when it tries to Popen() that `path`, it apparently fails. Is there somewhere else that uses the Python variable (`W3MIMGDISPLAY_PATH`) directly instead of `path`, perhaps? I couldn't see anything that looked relevant.

**Comment by hut** [comment]
> if I do an echo $W3MIMGDISPLAY at a bash prompt

Please check again that you wrote `export W3MIMGDISPLAY_PATH=/usr/libexec/w3m/w3mimgdisplay`, **with** the `_PATH`. I did exactly that mistake while debugging, and now it works for me.

**Comment by lycurgus** [comment]
>Please check again that you wrote `export W3MIMGDISPLAY_PATH`

Good catch - I'm 99.9% sure that was just a typo in the comment, but I'm not at the machine currently. I believe I included the `_PATH` in the variable name; `img_display.py` seemed to be picking it up correctly. It will be about 7 hours before I'm able to check that `export`, unfortunately - will advise when I've confirmed it's correct.

**Comment by lycurgus** [comment]
Ok, I've checked and it is indeed `W3MIMGDISPLAY_PATH` - the only difference between my line and yours is that I have single-quotes around the path itself; but those were added in as an earlier attempt at having the path recognised - the behaviour is the same with or without the quotes.

I also added in a second log print between the getting of the environment variable and the check for None in `img_display.py`.. the environment variable is definitely being read in correctly.

**Comment by hut** [comment]
This makes no sense to me. Just to make sure, you did all the other steps in https://github.com/hut/ranger/wiki/Image-Previews , right?

**Comment by lycurgus** [comment]
Yep, I've:
* got `w3m` and `w3m-img`
* copied config files in, via `ranger --copy-config=all` (just to be sure i have just removed and re-added all 4 files and re-edited rc.conf to set "preview_images" to true
* checked the other requirements - I'm on `urxvt` and a compatible version of ranger (1.6.1)

This was a clean install of ranger (and everything else, actually) so there shouldn't be any issues around upgrades.

At this point I'm wondering if it's some weird bug in Python - It's reading the environment variable just fine, but only works if the pre-existing variable is correct. I've tried removing the environment variable and there's no change in the behaviour (though the envvar is then properly returned as None).

I suppose my next step at this point is to build from source and see if it was just a problem with 1.6 somehow. Will advise how that goes...

**Comment by hut** [comment]
I'd suggest to go through the code with a python debugger like winpdb, or throw in some more prints/logging commands, to analyze what part exactly fails.

E.g. `raise Exception(repr(os.environ))` at the start of `ranger.ext.img_display.W3MImageDisplayer.initialize`. (Note: you need to run ranger with `--debug` to make that exception visible. Or press `W` for the log, but it's clunky.)

**Comment by lycurgus** [comment]
I used pdb to step into the code and confirmed that the environment variable is being read correctly - `subprocess.Popen()` was called successfully, and I could get the PID of the resulting process and `strace` it to watch it read the file and return the correct dimensions. The crash only occurred later than that.

After a bit of fruitless digging I hit upon the idea of grepping for the string "w3m" in the site-packages/ranger directory - which is how I noticed that gui/widgets/pager.py was also calling w3mimgdisplay.. but instead of using the environment variable, it relies (or relied) exclusively on img_display.W3MIMGDISPLAY_PATH. This explains why the previews worked only when the variable was changed - even though _w3mimgdisplay() in img_display.py was correctly getting and using the environment variable.

Looking at the current code (I also built and tried it out in a live USB environment) it seems changes to pager.py mean this is no longer an issue, so I'm happy to close it. I'll talk to the maintainers of the Fedora package to see about getting a more recent release packaged.

Thanks for your time on this one! I've learned a couple of things about debugging on the way :)
